### PR TITLE
Fix recovering from malformed compiler list

### PR DESCRIPTION
**Pull request by kanej**
This PR is an attempt to deal with #1466 

A halt during compiler download can lead the compiler list file to be partially written (and so invalid json). On a subsequent run the parsing of the compiler list fails and leaves the hardhat install in a broken state (requiring a `npx hardhat clean --global`).

This PR changes the compiler downloader behaviour to parse the file and on `SyntaxError`, redownload and retry the parse once more. Other errors that are not syntax based are rethrown.

## Manual testing

Additional tests have been added to the test suite, but if you are attempting to reproduce the bug manually:

1. create a new hardhat project based on the basic template
2. Start `npx hardhat node`
3. Run the sample deploy script `npx hardhat run ./scripts/sample-script.js --network localhost` triggering a compile (and compiler download)
4. Edit the `/Users/<user-name>/Library/Caches/hardhat-nodejs/compilers/macosx-amd64/list.json` file to be a partial and invalid json file, so deleting half of it for instance
5. Edit the `contracts/Greeter.sol` to force recompile on next deploy
6. Rerun the sample deploy script `npx hardhat run ./scripts/sample-script.js --network localhost`

That gives the expected error:

```shell
An unexpected error occurred:

SyntaxError: /Users/<user-name>/Library/Caches/hardhat-nodejs/compilers/macosx-amd64/list.json: Unexpected end of JSON input
    at JSON.parse (<anonymous>)
    at /Users/<user-name>/projects/sandbox/another/node_modules/jsonfile/index.js:33:18
    at /Users/<user-name>/projects/sandbox/another/node_modules/graceful-fs/graceful-fs.js:123:16
    at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:71:3)
```

With the fixes in this branch the second run of the deploy script will succeed, as it overwrites the compilers list json file on detecting it if malformed.

### CODE REVIEW

File: .changeset/strange-rocks-poke.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+"hardhat": patch
```
Comment (kanej) [inline .changeset/strange-rocks-poke.md:2]: Should this be `@hardhat-core`? @fvictorio

File: packages/hardhat-core/src/internal/solidity/compiler/downloader.ts [hunk 2]
```diff
@@ -159,7 +159,19 @@ export class CompilerDownloader {
       await this.downloadCompilersList(platform);
     }
 
-    return fsExtra.readJson(this.getCompilersListPath(platform));
+    try {
+      return await fsExtra.readJSON(this.getCompilersListPath(platform));
+    } catch (error) {
+      // if parsing throws a syntax error, redownload and parse once more
+      if (!(error instanceof SyntaxError)) {
+        // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
```
Comment (kanej) [inline packages/hardhat-core/src/internal/solidity/compiler/downloader.ts:167]: My intention here was to touch nothing unless it is explicitly the parsing error, so rethrowing ... but that does mean running against the eslint rule.

File: .changeset/strange-rocks-poke.md (hunk 1 above)
Comment (alcuadrado) [reply .changeset/strange-rocks-poke.md:2]: No, `hardhat` is fine. `packages/hardhat-core` is named like that to make talking about `hardhat` (the npm package) simpler

File: .changeset/strange-rocks-poke.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
...
+---
+
+Fix - allow recovering from partially downloaded compiler list
```
Comment (alcuadrado) [inline .changeset/strange-rocks-poke.md:5]: These messages end up in the changelong and other user-facing places, so we don't use any kind of structured message, but rather a plain English explanation. Something like 

> Fix a bug that prevented Hardhat from working if the compilers list was partially downloaded

File: packages/hardhat-core/src/internal/solidity/compiler/downloader.ts (hunk 2 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/solidity/compiler/downloader.ts:167]: I like this change!

I also think that we should implement the log to download to a tmp file and move once ready, as a general helper, because we download other things that are harder to validate. For example, the compiler's binaries.

Comment (kanej) [reply packages/hardhat-core/src/internal/solidity/compiler/downloader.ts:167]: I have modified `packages/hardhat-core/src/internal/util/download.ts` to download to a temp file in the same directory are the intended final filepath so `compilers/list.json` will initially be downloaded to `compilers/tmp-list.json`; on successful download the file is then moved into the intended final filepath.

I can only find two usages of download - download the compiler list and download an individual compiler. So I have made the tmp file change directly in `download.ts` but I can change that to be a separate function `download-via-tmp` or whatever and apply it on compilers list initially. What is your preference?

File: packages/hardhat-core/src/internal/solidity/compiler/downloader.ts [hunk 4]
```diff
@@ -159,7 +159,19 @@ export class CompilerDownloader {
...
+        throw error;
+      }
+
+      await this.downloadCompilersList(platform);
+
+      return await fsExtra.readJSON(this.getCompilersListPath(platform));
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/solidity/compiler/downloader.ts:173]: One alternative here is to do just `return getCompilersList(platform)`. That feels better to me, but if we do that we should also include a `retryLimit` variable.

Comment (fvictorio) [reply packages/hardhat-core/src/internal/solidity/compiler/downloader.ts:173]: Something like this:

```typescript
  public async getCompilersList(
    platform: CompilerPlatform,
    pendingRetries = 3
  ): Promise<CompilersList> {
    if (!(await this.compilersListExists(platform))) {
      await this.downloadCompilersList(platform);
    }

    try {
      return await fsExtra.readJSON(this.getCompilersListPath(platform));
    } catch (error) {
      // if parsing throws a syntax error, redownload and parse once more
      if (!(error instanceof SyntaxError) || pendingRetries === 0) {
        // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
        throw error;
      }

      // remove the malformed list and retry
      await fsExtra.remove(this.getCompilersListPath(platform));
      return this.getCompilersList(platform, pendingRetries - 1);
    }
  }
```

File: packages/hardhat-core/src/internal/solidity/compiler/downloader.ts (hunk 2 above)
Comment (alcuadrado) [reply packages/hardhat-core/src/internal/solidity/compiler/downloader.ts:167]: I think this makes sense

File: packages/hardhat-core/test/internal/solidity/compiler/downloader.ts [hunk 5]
```diff
@@ -316,6 +316,98 @@ describe("Compiler downloader", function () {
...
+              downloadCount++;
+            },
+            forceSolcJs: true,
+          });
+
+          try {
+            await failingDownloader.getCompilersList(CompilerPlatform.WASM);
+          } catch (error) {
+            assert.isTrue(error instanceof SyntaxError);
+          }
```
Comment (fvictorio) [inline packages/hardhat-core/test/internal/solidity/compiler/downloader.ts:359]: This doesn't quite work because this doesn't check that there is actually an error. You can use `isRejected` for this, see for example:

https://github.com/nomiclabs/hardhat/blob/f911aa3b76a61835a1a83740065370fc6ab79183/packages/hardhat-core/test/internal/core/providers/accounts.ts#L412-L418

File: packages/hardhat-core/test/internal/solidity/compiler/downloader.ts [hunk 6]
```diff
@@ -316,6 +316,98 @@ describe("Compiler downloader", function () {
         });
       });
     });
+
+    describe("When downloading fails", function () {
+      describe("getCompilersList", function () {
+        it("retries on a failed download", async function () {
+          let downloadCount = 0;
+          const failingDownloader = (mockDownloader = new CompilerDownloader(
```
Comment (fvictorio) [inline packages/hardhat-core/test/internal/solidity/compiler/downloader.ts:324]: This feels a bit weird. Can't you just assign to `mockDownloader` and use that?

File: packages/hardhat-core/test/internal/solidity/compiler/downloader.ts (hunk 5 above)
Comment (kanej) [reply packages/hardhat-core/test/internal/solidity/compiler/downloader.ts]: Thanks for that. I couldn't find the async throws assertion on `chai`, I hadn't grasped we had `chai-as-promised`; and `isRejected` actually works rather than my homebrew.

File: packages/hardhat-core/test/internal/solidity/compiler/downloader.ts (hunk 6 above)
Comment (kanej) [reply packages/hardhat-core/test/internal/solidity/compiler/downloader.ts]: I have moved the more complicated download function up, and used it across the tests in this block.

File: .changeset/wise-nails-lick.md [hunk 7]
```diff
@@ -2,4 +2,4 @@
 "@nomiclabs/hardhat-ethers": patch
 ---
 
-Adds a custom formatter to better display BigNumber's in Hardhat console (issue #2109).
+Adds a custom formatter to better display BigNumber's in Hardhat console (issue #2109)
```
Comment (fvictorio) [inline .changeset/wise-nails-lick.md:5]: This doesn't belong to this PR, does it?

(If this just means removing a period from the changeset for consistency or something, I don't mind. Just pointing it out in case it means something strange happened during the rebase.)

**Review by fvictorio — APPROVED** [review]
Pre-approving, just one very minor comment.

File: .changeset/wise-nails-lick.md (hunk 7 above)
Comment (kanej) [reply .changeset/wise-nails-lick.md:5]: You are correct, this shouldn't be here. I added in this changeset, wondered whether the convention was adding a fullstop, checked the other changesets and found one with a comma and one without and make both of mine consistent with the third.

Comment (kanej) [reply .changeset/wise-nails-lick.md]: The changeset has already been released as it turns out.

File: packages/hardhat-core/src/internal/util/download.ts [hunk 8]
```diff
@@ -9,6 +9,19 @@ interface FetchOptions {
   agent?: undefined | HttpsProxyAgent.HttpsProxyAgent;
 }
 
+const TEMP_FILE_PREFIX = "tmp-";
+
+function resolveTempFileName(filePath: string): string {
+  const { root, dir, ext, name } = path.parse(filePath);
+
+  return path.format({
+    root,
```
Comment (fvictorio) [inline packages/hardhat-core/src/internal/util/download.ts:18]: Small comment I missed the previous time I reviewed this:

According to [the docs](https://nodejs.org/api/path.html#pathformatpathobject), «pathObject.root is ignored if pathObject.dir is provided», so I think we can remove `root` from here and from the result of `path.parse`.

**Review by alcuadrado — APPROVED** [review]

File: packages/hardhat-core/src/internal/util/download.ts (hunk 8 above)
Comment (kanej) [reply packages/hardhat-core/src/internal/util/download.ts]: Fix applied
